src/Parser/DestructionWarlock/Modules/SoulShards/SoulShardEvents.js

```diff
@@ -0,0 +1,254 @@
+    this.owner.triggerEvent('soulshardfragment_gained', shardEvent);
+  }
+
+  //should also work with Havoc and talent Fire and Brimstone
+  on_byPlayer_damage(event) {
+    const spellId = event.ability.guid;
+    if (FRAGMENT_GENERATING_ABILITIES.indexOf(spellId) === -1) {
+      return;
+    }
+
+    const shardEvent = {
+      timestamp: event.timestamp,
+      type: 'soulshardfragment_gained',
+      ability: {
+        guid: spellId,
+        name: SPELLS[spellId].name,
+      },
+      targetID: event.targetID,
+      targetInstance: event.targetInstance,
+    };
+    let rawGain;
+    let gain = 0;
+    let waste = 0;
+
+    switch (spellId) {
+      case SPELLS.IMMOLATE_DEBUFF.id:
+        rawGain = 1;
+        break;
+      case SPELLS.CONFLAGRATE.id:
+        rawGain = 5;
+        break;
+      case SPELLS.INCINERATE.id:
+        const enemy = this.enemies.getEntity(event);
+        if (!enemy) {
+          //shouldn't happen, bail out
+          return;
+        }
+
+        const hasHavoc = enemy.hasBuff(SPELLS.HAVOC.id, event.timestamp);
+        //Havoc is somehow bugged in the sense that it doesn't gain the benefit of T20 2p set bonus, so if the target has Havoc, it doesn't matter if we have the set or not, otherwise it counts it in
+        rawGain = hasHavoc ? 2 : (this.hasT20_2p ? 3 : 2);
+        if (event.hitType === HIT_TYPES.CRIT) {
+          rawGain++;
+        }
+        break;
+      default:
+        break;
+    }
+    if (this.currentFragments + rawGain > MAX_FRAGMENTS) {
+      gain = MAX_FRAGMENTS - this.currentFragments;
+      waste = this.currentFragments + rawGain - MAX_FRAGMENTS;
+    }
+    else {
+      gain = rawGain;
+    }
+
+    this.currentFragments += gain;
+
+    shardEvent.amount = gain;
+    shardEvent.waste = waste;
+    shardEvent.currentFragments = this.currentFragments; //AFTER adding the amount
+
+    debug && console.log('++ ' + shardEvent.amount + '(w: ' + shardEvent.waste + ') = ' + shardEvent.currentFragments + ', ' + shardEvent.ability.name + ', orig: ', event);
+    this.owner.triggerEvent('soulshardfragment_gained', shardEvent);
+  }
+
+  on_byPlayer_cast(event) {
+    const spellId = event.ability.guid;
+    if (spellId === SPELLS.DIMENSIONAL_RIFT_CAST.id) {
+      this.processDimensionalRift(event);
+    }
+    else if (FRAGMENT_SPENDING_ABILITIES[spellId]) {
+      this.processSpenders(event);
+    }
+  }
+
+  //in order to de-clutter on_byPlayer_cast()
+  processDimensionalRift(event) {
+    const spellId = event.ability.guid;
+    const shardEvent = {
+      timestamp: event.timestamp,
+      type: 'soulshardfragment_gained',
+      ability: {
+        guid: spellId,
+        name: SPELLS[spellId].name,
+      },
+      targetID: event.targetID,
+      targetInstance: event.targetInstance,
+    };
+    const rawGain = 3;
+    let gain = 0;
```

Comment:
Seems like a lot of repeating logic here from the previous method, maybe refactor this to something more generic?

Reply:
I'll try to pull it out into processGenerators() or something like that, I agree that a lot is repeating there.